src/puya/ir/builder/state.py

```diff
@@ -0,0 +1,217 @@
+from puya.avm_type import AVMType
+from puya.awst import nodes as awst_nodes
+from puya.ir.avm_ops import AVMOp
+from puya.ir.builder.utils import assign_targets, mktemp
+from puya.ir.context import IRFunctionBuildContext
+from puya.ir.models import BytesConstant, Intrinsic, UInt64Constant, Value, ValueProvider
+from puya.ir.types_ import bytes_enc_to_avm_bytes_enc, wtype_to_avm_type
+
+
+def visit_app_state_expression(
+    context: IRFunctionBuildContext, expr: awst_nodes.AppStateExpression
+) -> ValueProvider:
+    state_def = context.resolve_state(expr.field_name, expr.source_location)
+    # TODO: add specific (unsafe) optimisation flag to allow skipping this check
+    current_app_offset = UInt64Constant(value=0, source_location=expr.source_location)
+    key = BytesConstant(
+        value=state_def.key,
+        source_location=expr.source_location,
+        encoding=bytes_enc_to_avm_bytes_enc(state_def.key_encoding),
+    )
+
+    # note: we manually construct temporary targets here since atype is any,
+    #       but we "know" the type from the expression
+    value_atype = wtype_to_avm_type(expr.wtype)
+    value_tmp = mktemp(
+        context,
+        atype=value_atype,
+        source_location=expr.source_location,
+        description="app_global_get_ex_value",
+    )
+    did_exist_tmp = mktemp(
+        context,
+        atype=AVMType.uint64,
+        source_location=expr.source_location,
+        description="app_global_get_ex_did_exist",
+    )
+    assign_targets(
+        context,
+        source=Intrinsic(
+            op=AVMOp.app_global_get_ex,
+            args=[current_app_offset, key],
+            source_location=expr.source_location,
+        ),
+        targets=[value_tmp, did_exist_tmp],
+        assignment_location=expr.source_location,
+    )
+    context.block_builder.add(
+        Intrinsic(
+            op=AVMOp.assert_,
+            args=[did_exist_tmp],
+            comment=f"check {expr.field_name} exists",
+            source_location=expr.source_location,
+        )
+    )
+
+    return value_tmp
```

Comment:
This code is a duplicate of `visit_checked_maybe`. Is it worth extracting a helper? 🤔

Reply:
also duplicated below

Reply:
Done some refactoring in the state file to extract some patterns here.
The code in `visit_checked_maybe` did fit the pattern but it didn't need to, so that's also been refactored.